agent/engine/dockerauth/ecr.go

```diff
@@ -17,54 +17,135 @@ import (
-const proxyEndpointScheme = "https://"
+const (
+	// MinimumJitterDuration is the minimum duration to mark the credentials
+	// as expired before it's actually expired
+	MinimumJitterDuration = 30 * time.Minute
+	roundtripTimeout      = 5 * time.Second
+	proxyEndpointScheme   = "https://"
+)
+
+// String formats the cachKey as a string
+func (key *cacheKey) String() string {
+	return fmt.Sprintf("%s-%s-%s-%s", key.roleARN, key.region, key.registryID, key.endpointOverride)
+}
 
 // NewECRAuthProvider returns a DockerAuthProvider that can handle retrieve
 // credentials for pulling from Amazon EC2 Container Registry
-func NewECRAuthProvider(authData *api.ECRAuthData, clientFactory ecr.ECRFactory) DockerAuthProvider {
-	if authData == nil {
-		return &ecrAuthProvider{}
-	}
-	log.Tracef("Getting client in %s with endpoint %s", authData.Region, authData.EndpointOverride)
+func NewECRAuthProvider(ecrFactory ecr.ECRFactory, cache async.Cache) DockerAuthProvider {
 	return &ecrAuthProvider{
-		authData: authData,
-		client:   clientFactory.GetClient(authData.Region, authData.EndpointOverride),
+		tokenCache: cache,
+		factory:    ecrFactory,
 	}
 }
 
 // GetAuthconfig retrieves the correct auth configuration for the given repository
-func (authProvider *ecrAuthProvider) GetAuthconfig(image string) (docker.AuthConfiguration, error) {
-	if authProvider.authData == nil {
-		return docker.AuthConfiguration{}, fmt.Errorf("ecrAuthProvider cannot be used without AuthData")
+func (authProvider *ecrAuthProvider) GetAuthconfig(image string,
+	authData *api.ECRAuthData) (docker.AuthConfiguration, error) {
+	if authData == nil {
+		return docker.AuthConfiguration{}, fmt.Errorf("ecr auth: missing container auth data")
 	}
+
+	// First try to get the token from cache, if the token does not exist,
+	// then call ECR api to get the new token
+	key := cacheKey{
+		region:           authData.Region,
+		endpointOverride: authData.EndpointOverride,
+		registryID:       authData.RegistryID,
+	}
+
+	// If the container is using execution role credentials to pull,
+	// add the roleARN as part of the cache key so that docker auth for
+	// containers pull with the same role can be cached
+	if authData.GetPullCredentials() != (credentials.IAMRoleCredentials{}) {
+		key.roleARN = authData.GetPullCredentials().RoleArn
+	}
+
+	// Try to get the auth config from cache
+	auth := authProvider.getAuthConfigFromCache(key)
+	if auth != nil {
+		return *auth, nil
+	}
+
+	// Get the auth config from ECR
+	return authProvider.getAuthConfigFromECR(image, key, authData)
+}
+
+// getAuthconfigFromCache retrieves the token from cache
+func (authProvider *ecrAuthProvider) getAuthConfigFromCache(key cacheKey) *docker.AuthConfiguration {
+	token, ok := authProvider.tokenCache.Get(key.String())
+	if !ok {
+		return nil
+	}
+
+	cachedToken, ok := token.(*ecrapi.AuthorizationData)
+	if !ok {
+		log.Warnf("Reading ECR credentials from cache failed")
+	} else if authProvider.IsTokenValid(cachedToken) {
```

Comment:
please return `nil` here so that we can get rid of the multiple level of indentations here.